client/allocrunner/consul_hook.go

```diff
@@ -196,25 +196,27 @@ func (h *consulHook) prepareConsulTokensForServices(services []*structs.Service,
 			continue
 		}
 
-		if err := h.getConsulTokens(clusterName, service.Identity.Name, tokens, req); err != nil {
-			mErr.Errors = append(mErr.Errors, err)
+		if merr := h.getConsulTokens(clusterName, service.Identity.Name, tokens, req); merr != nil {
+			mErr.Errors = append(mErr.Errors, merr.Errors...)
 			continue
 		}
 	}
 
-	return mErr.ErrorOrNil()
+	return mErr
 }
 
-func (h *consulHook) getConsulTokens(cluster, identityName string, tokens map[string]map[string]*consulapi.ACLToken, req map[string]consul.JWTLoginRequest) error {
+func (h *consulHook) getConsulTokens(cluster, identityName string, tokens map[string]map[string]*consulapi.ACLToken, req map[string]consul.JWTLoginRequest) *multierror.Error {
 	client, err := h.clientForCluster(cluster)
 	if err != nil {
-		return err
+		return &multierror.Error{Errors: []error{err}}
 	}
 
 	// get consul acl tokens
 	t, err := client.DeriveTokenWithJWT(req)
 	if err != nil {
-		return err
+		// DeriveTokenWithJWT actually returns a multierror, so we assert the
+		// type to make error formatting from the hook level more readable.
+		return err.(*multierror.Error)
 	}
```

Comment:
Oh, there are more layers 😅 

This gives me some pause because it makes an implicit assumption from the caller about an implementation detail. It's also tricky because `DeriveTokenWithJWT` comes from an interface, where another implementation may not return a multierror (a test mock for example).

Looking at the code again, I think the core problem stems from the use of `mErr.Errors = append(mErr.Errors, err)` everywhere. Appending an error directly like that creates a new level if the error being appended is a multierror since the top-level error is the ` 1 error occurred:` message.

What we usually want is to accumulate errors. For that you should use `mErr = multierror.Append(mErr, err)`.

https://github.com/hashicorp/nomad/commit/16d6cfaad00d43ced253a3862a2f46c65fb985cd make this change and I think works as expected

<img width="1208" alt="image" src="https://github.com/hashicorp/nomad/assets/775380/d2a236b6-9fa6-4bce-86ad-f5c2a8e7b33e">

I also noticed that the error from `h.revokeTokens()` was being ignored. Not sure if it was done on purpose but I bundled it there with the results.

How does that look to you?

Reply:
> What we usually want is to accumulate errors. For that you should use mErr = multierror.Append(mErr, err).

ooooh so _that's_ the mistake I made. I'll merge your branch into this.